src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -268,43 +333,60 @@ public NetworkPeer(IPEndPoint peerEndPoint, Network network, NetworkPeerConnecti
-            : this(true, peerEndPoint, network, parameters, dateTimeProvider, loggerFactory)
+        /// <param name="onDisconnected">Callback that is invoked when peer has finished disconnecting, or <c>null</c> when no notification after the disconnection is required.</param>
+        public NetworkPeer(IPEndPoint peerEndPoint,
+            Network network,
+            NetworkPeerConnectionParameters parameters,
+            TcpClient client,
+            IDateTimeProvider dateTimeProvider,
+            INetworkPeerFactory networkPeerFactory,
+            ILoggerFactory loggerFactory,
+            Action<INetworkPeer> onDisconnected = null)
+            : this(true, peerEndPoint, network, parameters, dateTimeProvider, loggerFactory, onDisconnected)
         {
             this.Connection = networkPeerFactory.CreateNetworkPeerConnection(this, client, this.ProcessMessageAsync);
 
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.Connection.Id}-{peerEndPoint}] ");
             this.logger.LogTrace("()");
 
-            this.RemoteSocketEndpoint = this.PeerEndPoint;
-            this.RemoteSocketAddress = this.RemoteSocketEndpoint.Address;
-            this.RemoteSocketPort = this.RemoteSocketEndpoint.Port;
-
             this.logger.LogTrace("Connected to peer '{0}'.", this.PeerEndPoint);
-            this.State = NetworkPeerState.Connected;
 
             this.InitDefaultBehaviors(this.ConnectionParameters);
             this.Connection.StartReceiveMessages();
 
             this.logger.LogTrace("(-)");
         }
 
-        /// <inheritdoc/>
-        public async Task SetStateAsync(NetworkPeerState newState)
+        /// <summary>
+        /// Sets a new network state of the peer.
+        /// </summary>
+        /// <param name="newState">New network state to be set.</param>
+        /// <remarks>This method is not thread safe.</remarks>
+        private async Task SetStateAsync(NetworkPeerState newState)
         {
             this.logger.LogTrace("({0}:{1},{2}:{3})", nameof(newState), newState, nameof(this.State), this.State);
 
             NetworkPeerState previous = this.State;
-            if (previous != newState)
+
+            if (stateTransitionTable[previous].Contains(newState))
             {
                 this.State = newState;
+
                 await this.OnStateChangedAsync(previous).ConfigureAwait(false);
 
                 if ((newState == NetworkPeerState.Failed) || (newState == NetworkPeerState.Offline))
+                {
                     this.logger.LogTrace("Communication with the peer has been closed.");
+
+                    this.ExecuteDisconnectedCallbackWhenSafe();
+                }
+            }
+            else if (previous != newState)
+            {
+                this.logger.LogDebug("Illegal transition from {0} to {1} occurred.", previous, newState);
```

Comment:
This should be much less forgiving it indicates a design error if we hit this `else`

Reply:
Agreed. Will change to Error level.

Reply:
Yes, but not always, which means that without any change just raising this to error level would produce fake errors. I believe during shutdown of the peer and its connection some components may try to set the new state to off-line while the state is already in failed state or vice versa, which seems perfectly valid to me in terms of that the transition should not happen but it should not be reported as an error.

Reply:
So if you want to change this, we should exclude transition from off-line to error and error to off-line.

Reply:
>Yes, but not always
I assumed any state transaction in the table is allowed and all others are invalid (and state can be changed only internally

Reply:
For that we would have to match off-line and failed states into one